### PR TITLE
Issue#155 implement synchronize my book

### PR DESCRIPTION
**Pull request by kstigen**
GlobalDigitalLibraryio/issues#155

### CODE REVIEW

File: src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala [hunk 1]
```diff
@@ -9,23 +9,118 @@ package no.gdl.bookapi.service.translation
...
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          })
+        }
+      }
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala]: You can remove the curly braces around the case bodies (4 pairs), for this to look a bit prettier :)

File: src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala [hunk 2]
```diff
@@ -9,23 +9,118 @@ package no.gdl.bookapi.service.translation
...
 
   class TranslationService extends LazyLogging {
-    def fetchTranslationsIfAllTranslated(projectIdentifier: String, language: LanguageTag): Try[Unit] = {
-      Success()
-      // TODO in #138: Implement fetch translations
+    def inTranslationWithId(inTranslationId: Long): Option[InTranslation] = {
+      translationDbService.translationWithId(inTranslationId)
+    }
+
+    def fetchUpdatesFor(inTranslation: InTranslation): Try[SynchronizeResponse] = {
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala]: Could this function be broken into multiple functions? I find it hard to read.

File: src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala [hunk 3]
```diff
@@ -9,23 +9,118 @@ package no.gdl.bookapi.service.translation
...
+                  }
+                }
+              }
+            }
+          })
+        }
+      }
+    }
+
+    def fetchTranslationsIfAllTranslated(inTranslationFile: InTranslationFile): Try[Unit] = {
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala]: Could this function be broken into multiple functions? I find it hard to read.

File: src/test/scala/no/gdl/bookapi/service/translation/MergeServiceTest.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  test("that img-tags are replaced with embed-tags") {
+    val original = TestData.Domain.DefaultChapter.copy(content = originalHtml)
+    val translated = TestData.Crowdin.DefaultTranslatedChapter.copy(content = translatedHtml)
+
+    val newChapter = service.mergeChapter(original, translated)
+    val newContentDoc = Jsoup.parseBodyFragment(newChapter.content)
+    newContentDoc.select("embed[data-resource]").size() should be (2)
+    newContentDoc.getElementsByTag("img").size() should be (0)
+  }
```
Comment (runarfu) [inline src/test/scala/no/gdl/bookapi/service/translation/MergeServiceTest.scala]: I would like to also assert that the order of images is kept intact. Suggested diff:
```
diff --git a/src/test/scala/no/gdl/bookapi/service/translation/MergeServiceTest.scala b/src/test/scala/no/gdl/bookapi/service/translation/MergeServiceTest.scala
index 760106c..6189781 100644
--- a/src/test/scala/no/gdl/bookapi/service/translation/MergeServiceTest.scala
+++ b/src/test/scala/no/gdl/bookapi/service/translation/MergeServiceTest.scala
@@ -19,6 +19,7 @@ class MergeServiceTest extends UnitSuite with TestEnvironment {
       |<p>
       | There was a large boulder on the grassy plain. He stumbled and fell down.
       |</p>
+      |<embed data-resource="image" data-resource_id="558"/>
       |""".stripMargin

   private val translatedHtml =
@@ -27,6 +28,7 @@ class MergeServiceTest extends UnitSuite with TestEnvironment {
       |<p>
       | Hjorten traff en stor stein på den grønne sletten. Han snublet og falt.
       |</p>
+      |<img src="https://images.test.digitallibrary.io/123.jpg" />
       |""".stripMargin

   test("that img-tags are replaced with embed-tags") {
@@ -35,7 +37,11 @@ class MergeServiceTest extends UnitSuite with TestEnvironment {

     val newChapter = service.mergeChapter(original, translated)
     val newContentDoc = Jsoup.parseBodyFragment(newChapter.content)
-    newContentDoc.select("embed[data-resource]").size() should be (2)
+    val embedElements = newContentDoc.select("embed[data-resource]")
+    embedElements.size() should be (3)
+    embedElements.get(0).toString should equal ("<embed data-resource=\"image\" data-resource_id=\"556\">")
+    embedElements.get(1).toString should equal ("<embed data-resource=\"image\" data-resource_id=\"557\">")
+    embedElements.get(2).toString should equal ("<embed data-resource=\"image\" data-resource_id=\"558\">")
     newContentDoc.getElementsByTag("img").size() should be (0)
   }

@@ -54,7 +60,7 @@ class MergeServiceTest extends UnitSuite with TestEnvironment {

     val newChapter = service.mergeChapter(original, translated)
     val newContentDoc = Jsoup.parseBodyFragment(newChapter.content)
-    newContentDoc.select("embed[data-resource]").size() should be (2)
+    newContentDoc.select("embed[data-resource]").size() should be (3)
     newContentDoc.getElementsByTag("img").size() should be (0)
   }
```

File: src/test/scala/no/gdl/bookapi/service/translation/TranslationServiceTest.scala [hunk 5]
```diff
@@ -154,4 +155,124 @@ class TranslationServiceTest extends UnitSuite with TestEnvironment {
     val result = service.updateTranslationStatus("abc", LanguageTag("nob"), "123", TranslationStatus.TRANSLATED)
     result should be a 'Success
   }
+
+  test("that fetchTranslationsIfAllTranslated returns Success when not all translated") {
```
Comment (runarfu) [inline src/test/scala/no/gdl/bookapi/service/translation/TranslationServiceTest.scala]: This appears as a contradiction to me. `fetchTranslationsIfAllTranslated` sounds like it shouldn't return a `Success`?

File: src/main/scala/no/gdl/bookapi/service/ConverterService.scala [hunk 6]
```diff
@@ -172,6 +183,20 @@ trait ConverterService {
...
+    def toApiMyBook(inTranslation: InTranslation, translation: domain.Translation, availableLanguages: Seq[LanguageTag], book: api.Book): api.MyBook = {
+      api.MyBook(
+        book.id,
+        book.revision,
+        translation.title,
+        translation.translatedFrom.map(toApiLanguage),
+        toApiLanguage(translation.language),
+        book.publisher,
+        toApiCoverPhoto(translation.coverphoto),
+        synchronizeUrl = s"${BookApiProperties.Domain}${BookApiProperties.TranslationsPath}/synchronized/${inTranslation.id.get}",
```
[the remark this answers is not in the document]
Comment (kstigen) [reply src/main/scala/no/gdl/bookapi/service/ConverterService.scala]: Will not fix

File: src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala [hunk 7]
```diff
@@ -201,6 +166,72 @@ trait TranslationService {
...
+    private def findMetadataFile(inTranslation: InTranslation, translationFiles: Seq[InTranslationFile]): Try[InTranslationFile] = {
+      translationFiles.find(_.fileType == FileType.METADATA) match {
+        case None => Failure(new RuntimeException(s"No metadata for translation with id ${inTranslation.id} found. Cannot fetch updates"))
+        case Some(metadataFile) => Success(metadataFile)
+      }
+    }
+
+    private def findChapterFiles(translation: InTranslation, translationFiles: Seq[InTranslationFile]): Try[Seq[InTranslationFile]] = {
+      translationFiles.filter(_.fileType == FileType.CONTENT) match {
+        case head :: tail => Success(head :: tail)
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala]: To do this similarly to the other functions, you could match on `case Nil => Failure(…` first and then you know that the list is nonempty in the other case.

File: src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala [hunk 8]
```diff
@@ -9,23 +9,83 @@ package no.gdl.bookapi.service.translation
...
   class TranslationService extends LazyLogging {
-    def fetchTranslationsIfAllTranslated(projectIdentifier: String, language: LanguageTag): Try[Unit] = {
-      Success()
-      // TODO in #138: Implement fetch translations
+    def allFilesTranslated(inTranslationFile: InTranslationFile): Boolean = {
+      val translationFiles = translationDbService.filesForTranslation(inTranslationFile.inTranslationId)
+      translationFiles.forall(_.translationStatus == TranslationStatus.TRANSLATED)
+    }
+
+    def fetchUpdatesFor(inTranslation: InTranslation): Try[SynchronizeResponse] = {
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/service/translation/TranslationService.scala]: Beautiful!

**Pull request merged**